R/verify.R

```diff
@@ -0,0 +1,52 @@
+#' Reproducibility Check for `@code` Slot in `teal_data`
+#'
+#' Verifies the reproducibility of the `teal_data@env` environment using the `@code` slot.
+#'
+#' @return If the objects returned by the `@code` slot in the input `teal_data` object exactly match
+#' the objects in `teal_data@env`, the function updates the `@verified` slot to `TRUE` in the `teal_data` object.
+#' If the `@code` fails to recreate objects in `teal_data@env`, an error is thrown.
+#'
+#' @param x `teal_data` object
+#' @examples
+#' tdata1 <- teal_data()
+#' tdata1 <- within(tdata1, {
+#'   a <- 1
+#'   b <- a^5
+#'   c <- list(x = 2)
+#' })
+#' verify(tdata1)
+#'
+#' tdata2 <- teal_data(x1 = iris, code = "x1 = iris")
+#' verify(tdata2)
+#' verify(tdata2)@verified
+#' tdata2@verified
+#'
+#' \dontrun{
+#' tdata3 <- teal_data()
+#' tdata3 <- within(tdata3, {
+#'   stop("error")
+#' })
+#' }
+#' @name verify
+#' @aliases verify,teal_data-method
+#' @aliases verify,qenv.error-method
+#' @rdname verify
+#' @export
+setGeneric("verify", function(x) standardGeneric("verify"))
+setMethod("verify", "teal_data", definition = function(x) {
+  if (x@verified) {
+    return(x)
+  }
+  new_teal_data <- eval_code(teal_data(), x@code)
+
+  reproducible <- all.equal(x@env, new_teal_data@env)
+  if (isTRUE(reproducible)) {
+    x@verified <- TRUE
+    x
+  } else {
+    stop("Code verification failed.")
+  }
+})
+setMethod("verify", "qenv.error", definition = function(x) {
+  stop(x$trace)
```

Comment:
```suggestion
  stop(x$message, call. = FALSE)
```

Reply:
Hey @chlebowa I decided to return `object@trace` the same as we return it for `get_code.qenv.error in teal.code` https://github.com/insightsengineering/teal.code/blob/main/R/qenv-get_code.R#L37-L48

Reply:
Do we need the entire trace to be displayed when verification fails, though?

Reply:
I'm not sure. I tried to be consistent for `get_code` run on the same object in two family packages.

Reply:
If the code is 50 lines long and ~~validation~~ fails, do I want to read the entire trace or just the error message (that may or may not contain the offending call)? Legitimate question.

Reply:
I know it legitimate. And I'm on your side. I would also propose a change in teal.code then. Whatchya think?

Reply:
Well, with `get_code` I would expect to get everything.

Reply:
> Well, with get_code I would expect to get everything.

Ok I'm confused. @message is less content, but you would expect to get more, so rather `@trace` (combined with `@message` if it does not exist in `@trace`)?

Reply:
`get_code` should show everything, hence `qenv.error$trace` (which is the entirety of `qenv@code`).
I think with `verify` the error message is quite enough, hence `qenv.error$message`.